--- video.c.orig	2012-10-31 20:10:09.000000000 +0100
+++ video.c	2012-11-05 11:30:13.673477746 +0100
@@ -7493,11 +7493,15 @@
     switch (Osd3DMode) {
 	case 1:
+	    output_rect.x0 = output_rect.x0 / 2;
 	    output_rect.x1 = output_rect.x1 / 2;
-	    output_double_rect.x0 = output_rect.x1;
+	    output_double_rect.x0 = output_rect.x0 + (VideoWindowWidth / 2);
+	    output_double_rect.x1 = output_rect.x1 + (VideoWindowWidth / 2);
 	    break;
 	case 2:
+	    output_rect.y0 = output_rect.y0 / 2;
 	    output_rect.y1 = output_rect.y1 / 2;
-	    output_double_rect.y0 = output_rect.y1;
+	    output_double_rect.y0 = output_rect.y0 + (VideoWindowHeight / 2);
+	    output_double_rect.y1 = output_rect.y1 + (VideoWindowHeight / 2);
 	    break;
 	default:
 	    break;
